server/src/test/java/org/elasticsearch/index/mapper/SourceFieldMapperTests.java

```diff
@@ -172,13 +188,38 @@ public void testSourceObjectContainsExtraTokens() throws Exception {
         assertThat(exception.getRootCause().getMessage(), containsString("Unexpected close marker '}'"));
     }
 
-    public void testSyntheticDisabledNotSupported() throws Exception {
+    public void testSyntheticDisabledNotSupported() {
         Exception e = expectThrows(
             MapperParsingException.class,
             () -> createDocumentMapper(
-                topMapping(b -> b.startObject("_source").field("enabled", false).field("synthetic", true).endObject())
+                topMapping(b -> b.startObject("_source").field("enabled", false).field("mode", "synthetic").endObject())
             )
         );
-        assertThat(e.getMessage(), containsString("_source may not be disabled when setting [synthetic: true]"));
+        assertThat(e.getMessage(), containsString("Cannot set both [mode] and [enabled] parameters"));
+    }
+
+    public void testSyntheticUpdates() throws Exception {
+        MapperService mapperService = createMapperService("""
+            { "_doc" : { "_source" : { "mode" : "synthetic" } } }
+            """);
+
+        SourceFieldMapper mapper = mapperService.documentMapper().sourceMapper();
+        assertTrue(mapper.enabled());
+        assertTrue(mapper.isSynthetic());
+
+        merge(mapperService, """
+            { "_doc" : { "_source" : { "mode" : "synthetic" } } }
+            """);
+        mapper = mapperService.documentMapper().sourceMapper();
+        assertTrue(mapper.enabled());
+        assertTrue(mapper.isSynthetic());
+
+        ParsedDocument doc = mapperService.documentMapper().parse(source("{}"));
+        assertNull(doc.rootDoc().get(SourceFieldMapper.NAME));
+
+        Exception e = expectThrows(IllegalArgumentException.class, () -> merge(mapperService, """
+            { "_doc" : { "_source" : { "mode" : "stored" } } }
+            """));
+        assertThat(e.getMessage(), containsString("Cannot update parameter [mode] from [synthetic] to [stored]"));
```

Comment:
do we also need to test other updates like going to DISABLED or from STORED to SYNTHETIC?

Reply:
There's already a test that going from STORED to SYNTHETIC works.  I'll add a check that you can always go to DISABLED.